libcrux-psq/src/handshake.rs

```diff
@@ -128,51 +196,92 @@ pub(super) fn derive_k0(
         g_xs: &DHSharedSecret::derive(own_sk, peer_pk)?,
     };
 
-    Ok((tx0, AEADKey::new(&ikm, &tx0)?))
+    Ok((tx0, AEADKeyNonce::new(&ikm, &tx0, aead_type)?))
 }
 
-// K1 = KDF(K0 | g^cs | SS, tx1)
-pub(super) fn derive_k1(
-    k0: &AEADKey,
+// K1 = KDF(K0 | g^cs | [SS], tx1)
+pub(super) fn derive_k1_dh(
+    k0: &AEADKeyNonce,
     own_longterm_key: &DHPrivateKey,
     peer_longterm_pk: &DHPublicKey,
     pq_shared_secret: Option<PQSharedSecret>,
     tx1: &Transcript,
-) -> Result<AEADKey, HandshakeError> {
+    aead_type: AEAD,
+) -> Result<AEADKeyNonce, HandshakeError> {
     #[derive(TlsSerializeBytes, TlsSize)]
-    struct K1Ikm<'a> {
-        k0: &'a AEADKey,
+    struct K1IkmDh<'a> {
+        k0: &'a AEADKeyNonce,
         ecdh_shared_secret: &'a DHSharedSecret,
         pq_shared_secret: Option<PQSharedSecret>,
     }
 
     let ecdh_shared_secret = DHSharedSecret::derive(own_longterm_key, peer_longterm_pk)?;
+    let ikm = K1IkmDh {
+        k0,
+        ecdh_shared_secret: &ecdh_shared_secret,
+        pq_shared_secret,
+    };
+
+    // XXX: This makes clippy unhappy, but is a lifetime error for feature `classic-mceliece` if we return directlyr
```

Comment:
Disable the lint here then.

```suggestion
    // XXX: This makes clippy unhappy, but is a lifetime error for feature `classic-mceliece` if we return directly
```